tools/gbagfx/jasc_pal.c

```diff
@@ -2,155 +2,87 @@
+        FATAL_ERROR("Invalid signature, expected \"JASC-PAL\", read: \"%s\"\n",line_buffer);
 
-    if (strcmp(line, "0100") != 0)
+    ReadJascPaletteLine(fp, line_buffer);
+    if (strcmp(line_buffer, "0100") != 0)
         FATAL_ERROR("Unsuported JASC-PAL version.\n");
 
-    ReadJascPaletteLine(fp, line);
+    // Get number of colors in palette
+    ReadJascPaletteLine(fp, line_buffer);
+    numColors = strtol(line_buffer, NULL, 10);
 
-    if (!ParseNumber(line, NULL, 10, &palette->numColors))
-        FATAL_ERROR("Failed to parse number of colors.\n");
-
-    if (palette->numColors < 1 || palette->numColors > 256)
-        FATAL_ERROR("%d is an invalid number of colors. The number of colors must be in the range [1, 256].\n", palette->numColors);
-
-    for (int i = 0; i < palette->numColors; i++)
+    // Check for sensible number of colors
+    if (numColors > 0 && numColors <= 256)
     {
-        ReadJascPaletteLine(fp, line);
-
-        char *s = line;
-        char *end;
-
-        int red;
-        int green;
-        int blue;
-
-        if (!ParseNumber(s, &end, 10, &red))
-            FATAL_ERROR("Failed to parse red color component.\n");
-
-        s = end;
-
-        if (*s != ' ')
-            FATAL_ERROR("Expected a space after red color component.\n");
-
-        s++;
-
-        if (*s < '0' || *s > '9')
-            FATAL_ERROR("Expected only a space between red and green color components.\n");
-
-        if (!ParseNumber(s, &end, 10, &green))
-            FATAL_ERROR("Failed to parse green color component.\n");
-
-        s = end;
-
-        if (*s != ' ')
-            FATAL_ERROR("Expected a space after green color component.\n");
-
-        s++;
-
-        if (*s < '0' || *s > '9')
-            FATAL_ERROR("Expected only a space between green and blue color components.\n");
-
-        if (!ParseNumber(s, &end, 10, &blue))
-            FATAL_ERROR("Failed to parse blue color component.\n");
-
-        if (*end != 0)
-            FATAL_ERROR("Garbage after blue color component.\n");
-
-        if (red < 0 || red > 255)
-            FATAL_ERROR("Red color component (%d) is outside the range [0, 255].\n", red);
-
-        if (green < 0 || green > 255)
-            FATAL_ERROR("Green color component (%d) is outside the range [0, 255].\n", green);
+        palette->numColors = numColors;
+    } else {
+        FATAL_ERROR("Out-of-Bounds number of colors specifed: %i. Maximum supported is 256\n", numColors);
+    }
 
-        if (blue < 0 || blue > 255)
-            FATAL_ERROR("Blue color component (%d) is outside the range [0, 255].\n", blue);
+    // Get color entries
+    for (int i = 0; i < numColors; ++i)
+    {
+        if (ReadJascPaletteLine(fp, line_buffer) == 0)
+            FATAL_ERROR("Failed to read color index %i\n", i);
+        if (sscanf(line_buffer, "%d %d %d", &red, &green, &blue) != 3)
+            FATAL_ERROR("Invalid color format in color \"%s\"\n", line_buffer);
```

Comment:
This is quietly accepting numbers out of the expected value range.

Reply:
great catch, thank you!